### PR TITLE
MRG: Compute patterns for multiple regressors

### PR DESCRIPTION
**Pull request by kingjr**
This enhancement allows retrieving multiple co-linear patterns in case `Y` is multidimensional (e.g. left/right audio/visual).

While doing this, I encountered the following issue: FWIU, the Haufe's pattern trick only works when X and Y are z-scored. Consequently, I am not sure whether we should force `LinearModel` to apply such z-scoring inside (and break backward compatibility), or whether we should just warn the user (which creates possibilities of doing it wrong). I originally thought we would be fine with `StandardScaler` prior to the LinearModel, but this won't suffice for ensuring Y is centered over 0 with std of 1.

Overall, I would like to get second pair of eye o the validity of my equations.

### CODE REVIEW

**Comment by wmvanvliet** [comment]
Why would the Haufe trick need X and y to be z-scored? The paper says they need to be zero-mean, and that is only so that the covariance can be computed as X.T.dot(X).

We should contact Haufe himself maybe.

**Comment by wmvanvliet** [comment]
Here, check it out:
https://gist.github.com/wmvanvliet/08f9b4b8c5e4f1bb0a25d5eceae1b419

**Comment by kingjr** [comment]
Thanks, for looking into it. You're right, we don't need to divide by the std, removing the mean is sufficient. 

I am still a bit confused with the `normalize` parameter in sklearn though. In you notebook, if you use a `Ridge` instead of a `LinearRegression` and set `normalize=True`, the patterns aren't correct anymore, right?

**Comment by wmvanvliet** [comment]
What `normalize=True` does is first this:

    X_scale = X.std(axis=0)
    X /= X_scale

and at the end of all things, this:

    self.coef_ *= X_scale

So, the weights are computed on normalized data, but afterwards re-scaled to work with the un-normalized data. Computing the weights on normalized data makes sure that each feature is treated equally, regardless of its scale. If scale of a feature is not correlated with its relevance, then this makes the computation more accurate, and also the computed patterns more accurate.

In my example in the Gist, I've disabled the noise, so the computation is 100% accurate regardless of normalization.

Now, in the case of Ridge regression where `alpha > 0`, you perform regularization of the covariance matrix. In the case of zero noise and enough data for the covariance matrix to be properly estimated, this will make the computation of the weights and hence the patterns less accurate. In the case of lots of noise (neurodata anyone?) or not enough data for the covariance matrix to be estimated, this will usually make the computation of the weights and patterns more accurate. Generally speaking, in regression, the more accurate the regression model, the more accurate the corresponding patterns.

File: mne/decoding/tests/test_base.py [hunk 1]
```diff
@@ -12,6 +13,95 @@
...
+    A = np.random.randn(n_features, n_targets)
+
+    X = Y.dot(A.T)
+    X += np.random.randn(n_samples, n_features)  # add noise
+    X += np.random.rand(n_features)  # Put an offset
+
+    return X, Y, A
+
+
+def __make_data(n_samples=1000, n_features=5, n_targets=3, noise_scale=2):
```
Comment (wmvanvliet) [inline mne/decoding/tests/test_base.py]: Yay, I was hoping my Gist code would end up as a unit test :).
But do we really need two functions that are called `make_data` and generate test data?

Comment (kingjr) [inline mne/decoding/tests/test_base.py]: oops, not properly cleaned up

**Comment by kingjr** [comment]
unrelated error ? @Eric89GXL

**Comment by kingjr** [comment]
ready for review

File: mne/decoding/tests/test_base.py [hunk 2]
```diff
@@ -12,6 +13,48 @@
...
+    """
+
+    # Define Y latent factors
+    np.random.seed(0)
+    cov_Y = np.eye(n_targets) * 10 + np.random.rand(n_targets, n_targets)
+    mean_Y = np.random.rand(n_targets)
+    Y = np.random.multivariate_normal(mean_Y, cov_Y, size=n_samples)
+
+    # Put a scale and offset to Y
+    Y += np.random.rand(n_targets)  # Put an offset
```
Comment (wmvanvliet) [inline mne/decoding/tests/test_base.py]: this line is actually not necessary. You already specify a non-zero mean on line 42.

File: mne/decoding/tests/test_base.py [hunk 3]
```diff
@@ -20,7 +63,9 @@ def test_get_coef():
     from sklearn.base import TransformerMixin, BaseEstimator
     from sklearn.pipeline import make_pipeline
     from sklearn.preprocessing import StandardScaler
-    from sklearn.linear_model import LinearRegression
+    from sklearn.linear_model import Ridge
+
+    lm = LinearModel(Ridge())
```
Comment (wmvanvliet) [inline mne/decoding/tests/test_base.py]: When using Ridge regression with `alpha=1`, you don't reconstruct the patterns exactly anymore. Why use Ridge instead of OLS regression?

File: mne/decoding/tests/test_base.py [hunk 4]
```diff
@@ -92,39 +133,35 @@ def inverse_transform(self, X):
...
+            assert_equal(patterns[0, 0], -patterns[0, 1])
     for t in [0, 1]:
         assert_array_equal(get_coef(clf.estimators_[t], 'filters_', False),
                            filters[:, t])
 
+    # Check patterns with more than 1 regressor
+    X, Y, A = _make_data(n_samples=2000, n_targets=2)
+    lm.fit(X, Y)
+    assert_array_equal(lm.filters_.shape, lm.patterns_.shape, [2, 5])
+    assert_array_equal(np.round(A * 10), np.round(lm.patterns_.T * 10))
```
Comment (wmvanvliet) [inline mne/decoding/tests/test_base.py]: The patterns found by the model are really not close to the actual patterns, are they?

File: mne/decoding/tests/test_base.py (hunk 2 above)
Comment (kingjr) [inline mne/decoding/tests/test_base.py]: right

File: mne/decoding/tests/test_base.py (hunk 3 above)
Comment (kingjr) [inline mne/decoding/tests/test_base.py]: I thought it would be useful to write down explicitly that the Ridge should get somewhat close but not exactly the result. You think we should just check for the LinearRegression case?

Comment (wmvanvliet) [inline mne/decoding/tests/test_base.py]: Ah, you check both cases. That's fine then.

File: mne/decoding/tests/test_base.py (hunk 4 above)
Comment (wmvanvliet) [inline mne/decoding/tests/test_base.py]: This should do the same, but clearer:

    assert_allclose(A, lm.patterns_.T, rtol=0.1)

File: mne/decoding/base.py [hunk 5]
```diff
@@ -83,25 +83,27 @@ def fit(self, X, y):
...
+        # Computes patterns using Haufe's trick:
+        # A = Covariance_X . W . Precision_Y
 
-            raise ValueError('model needs a unidimensional coef_ attribute to '
-                             'compute the patterns')
-        self.filters_ = np.squeeze(self.model.coef_)
-        self.patterns_ = np.cov(X.T).dot(self.filters_)
+        inv_Y = 1.
+        X = X - X.mean(0, keepdims=True)
+        if np.squeeze(y).ndim > 1:
```
Comment (Eric89GXL) [inline mne/decoding/base.py]: `squeeze` is generally unsafe and usually ambiguous. Can you check `ndim` or `shape` explicitly somehow to achieve the same thing more explicitly?

File: mne/decoding/base.py [hunk 6]
```diff
@@ -83,25 +83,27 @@ def fit(self, X, y):
...
+            inv_Y = np.linalg.pinv(np.cov(y.T))
+        self.patterns_ = np.cov(X.T).dot(self.filters_.T.dot(inv_Y)).T
 
         return self
 
+    @property
+    def filters_(self):
+        if not hasattr(self.model, 'coef_'):
+            raise ValueError('model does not have a `coef_` attribute.')
+        return np.squeeze(self.model.coef_)
```
Comment (Eric89GXL) [inline mne/decoding/base.py]: Docstring does not suggest that these can be 1D (but the squeeze can make it so)

File: mne/decoding/tests/test_base.py [hunk 7]
```diff
@@ -92,39 +130,37 @@ def inverse_transform(self, X):
...
-                               [n_features, n_times])
+                               X.shape[1:])
+            # the two time samples get inverted patterns
+            assert_equal(patterns[0, 0], -patterns[0, 1])
     for t in [0, 1]:
         assert_array_equal(get_coef(clf.estimators_[t], 'filters_', False),
                            filters[:, t])
 
+    # Check patterns with more than 1 regressor
+    X, Y, A = _make_data(n_samples=5000, n_features=5, n_targets=3)
```
Comment (Eric89GXL) [inline mne/decoding/tests/test_base.py]: Are there tests for `n_features=1`? This might help show `squeeze` giving unexpected results.

File: mne/decoding/base.py (hunk 5 above)
Comment (mmagnuski) [inline mne/decoding/base.py]: You can also specify dimension to squeeze (`axis`) which makes it safer and clearer :)

**Comment by kingjr** [comment]
Yes I know but the inconsistency comes from sklearn.perhaps @agramfort has
an idea on the cases where linear models give a y_pred of shape (n_samples,
1) versus (n_samples,)

**Comment by Eric89GXL** [comment]
Yes from a recent PR:

https://github.com/mne-tools/mne-python/pull/4315#issuecomment-308951075

**Comment by Eric89GXL** [comment]
... although you're asking about prediction, hopefully you can see it [from the code](https://github.com/mne-tools/mne-python/pull/4315/files#diff-61dcf21bf41aee6f6e49ca4eabc0b034R228), namely that it depends on the `fit` shape.

**Comment by kingjr** [comment]
Just for future reference, the size variation of y_predict depends on 1) the model and 2) the shape of y:

```python
import numpy as np
from sklearn.linear_model import Ridge, LinearRegression, LogisticRegression
from sklearn.svm import LinearSVC, LinearSVR
X = np.random.randn(10, 2)
y = np.random.randint(0, 2, 10)
for this_y in (y, y[:, None]):
    print('\ny.ndim: %s' % this_y.ndim)
    for clf in (LinearRegression(), LogisticRegression(), Ridge(), LinearSVC(), LinearSVR()):
        name = str(clf.__class__).split('.')[-1][:-2]
        print(name + ': ' + str(clf.fit(X, this_y).predict(X).shape))
```

```
y.ndim: 1
LinearRegression: (10,)
LogisticRegression: (10,)
Ridge: (10,)
LinearSVC: (10,)
LinearSVR: (10,)

y.ndim: 2
LinearRegression: (10, 1)
LogisticRegression: (10,)
Ridge: (10, 1)
LinearSVC: (10,)
LinearSVR: (10,)

```

**Comment by wmvanvliet** [comment]
I'm in favor, for our code, to always have `y.shape == y_hat.shape`

**Comment by kingjr** [comment]
I don't want to depart from sklearn, I think we should output whatever
sklearn outputs.

**Comment by Eric89GXL** [comment]
> I think we should output whatever sklearn outputs.

But it's inconsistent across models, so we can choose to do what Ridge/LinearRegression do and have `y.shape == y_predict.shape` (at least for 1D and `(N, 1)` arrays), right?

**Comment by kingjr** [comment]
I don't see why should impose a different behavior to sklearn. 

If users switch from `LinearSVR()` to `LinearModel(LinearSVR())`, then the dimensionality of the y_pred shouldn't change. The role of the class is just to provide a wrapper to internally compute `patterns_` and `filters_`.

**Comment by Eric89GXL** [comment]
> I don't see why should impose a different behavior to sklearn.

I agree, but sklearn itself is inconsistent, right? So we have to pick one? What you said before was this:

> y.ndim: 2
> LinearRegression: (10, 1)
> LogisticRegression: (10,)
> Ridge: (10, 1)
> LinearSVC: (10,)
> LinearSVR: (10,)

In other words, LinearRegression and Ridge do one thing (what I would argue is the right thing), while the other three do something else. Am I missing something?

**Comment by agramfort** [comment]
well sklearn should not be inconsistent. Can you provide gist to report
this?

**Comment by kingjr** [comment]
already discussed here: https://github.com/scikit-learn/scikit-learn/issues/5058

**Comment by kingjr** [comment]
@Eric89GXL this doesn't affect the prediction or patterns computations. I don't think we need to fix this.

File: mne/decoding/base.py [hunk 8]
```diff
@@ -71,37 +71,42 @@ def fit(self, X, y):
         ----------
         X : array, shape (n_samples, n_features)
             The training input samples to estimate the linear coefficients.
-        y : array, shape (n_samples,)
+        y : array, shape (n_samples,) or (n_samples, n_targets)
```
Comment (Eric89GXL) [inline mne/decoding/base.py]: more compact and consistent as `shape (n_samples,[ n_targets])` to show optional

File: mne/decoding/tests/test_base.py (hunk 7 above)
Comment (Eric89GXL) [inline mne/decoding/tests/test_base.py]: I still don't see any `n_features=1`, is it not worth testing? There is a remaining `np.squeeze` in the code.

**Comment by kingjr** [comment]
Comments addressed. 

There may be a remaining disagreement on the shape of y_pred; I'm in favour of not departing from sklearn (i.e. status quo), Eric and Marijn seem to favour a consistent y_pred.shape = y.shape. I suggest we move this to a separate Issue/PR, since the present one is on the computation of patterns.

**Comment by agramfort** [comment]
thx @kingjr
